substrate/frame/transaction-payment/src/lib.rs

```diff
@@ -943,3 +959,129 @@ where
 		Self::compute_actual_fee(len, &info, &post_info, Zero::zero())
 	}
 }
+
+use codec::Codec;
+use core::fmt::Debug;
+use frame_support::{CloneNoBound, EqNoBound, PartialEqNoBound, RuntimeDebugNoBound};
+use scale_info::StaticTypeInfo;
+use sp_io::hashing::blake2_256;
+use sp_runtime::{
+	impl_tx_ext_default,
+	traits::{IdentifyAccount, Verify},
+};
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum FeeAgent<AccountId> {
+	Origin,
+	AnyAgent,
+	Agent(AccountId),
+}
+
+impl<AccountId> Default for FeeAgent<AccountId> {
+	fn default() -> Self {
+		Self::Origin
+	}
+}
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct Context<AccountId> {
+	fee_agent: Option<AccountId>,
+}
+
+impl<AccountId> Default for Context<AccountId> {
+	fn default() -> Self {
+		Self { fee_agent: None }
+	}
+}
+
+/// Transaction extension that sets the transaction fee agent, the account sponsoring the fee, to
+/// the given account ID if the provided signature by that account is valid for all subsequent
+/// extensions. If signature is not provided, this extension is no-op.
+// TODO better doc.
+#[derive(
+	CloneNoBound, EqNoBound, PartialEqNoBound, Encode, Decode, RuntimeDebugNoBound, TypeInfo,
+)]
+#[codec(encode_bound())]
+#[codec(decode_bound())]
+pub struct SetFeeAgent<V: Verify>
+where
+	V: Codec + Debug + Sync + Send + Clone + Eq + PartialEq + StaticTypeInfo,
```

Comment:
better to avoid all those duplicated `Codec + Debug + Sync + Send + Clone + Eq + PartialEq + StaticTypeInfo`